content/flux/v0.x/function-type-signatures.md

```diff
@@ -0,0 +1,188 @@
+// explicit integer type.
+stream[{A with count: int}]
+```
+
+### Basic types
+
+Type signatures identify [basic types](/flux/v0.x/data-types/basic/) with the
+following type identifiers:
+
+```js
+bool     // boolean type
+bytes    // bytes type
+duration // duration type
+float    // float type
+int      // integer type
+string   // string type
+time     // time type
+uint     // unsigned integer type
+```
+
+### Composite types
+
+Type signatures identify Flux [composite types](/flux/v0.x/data-types/composite/)
+with the following syntaxes:
+
+```js
+[A]             // array type
+[B: A]          // dictionary type
+(param: A) => B // function type
+{}              // record type
+```
+
+### Regular expression types
+```js
+regexp // regular expression type
+```
+
+## Parameter notation
+
+Parameter notation indicates specific behaviors of function parameters.
+
+```js
+?  // Optional parameter
+<- // Pipe receive – indicates the parameter, by default, represents
+   // the piped-forward value or stream of tables
+```
+
+## Type constraints
+
+Some function parameters are "polymorphic" and can support multiple data types.
+Polymorphic parameters are bound by **type constraints**, which define what
+types can be used.
+Type signatures indicate type contraints for specific values using the
+`where A: Constraint` syntax.
+
+For more information about the different type constraints and the types each
+supports, see [Type constraints](/flux/v0.x/spec/types/#type-constraints).
+
+```js
+(v: A) => int where A: Timeable
+```
+
+## Example function type signatures
+
+- [Function without parameters](#function-without-parameters)
+- [Function with parameters](#function-with-parameters)
+- [Pass-through transformation](#pass-through-transformation)
+- [Basic transformation](#basic-transformation)
+- [Transformation that adds a column with an explicit type](#transformation-that-adds-a-column-with-an-explicit-type)
+
+---
+
+#### Function without parameters
+The following type signature describes a function that:
+
+- Has no parameters
+- Returns a time value
+
+```js
+() => time
+```
+
+---
+
+#### Function with parameters
+The following type signature describes a function that:
+
+- Has two parameters:
+  - multiplier _(Optional)_
+  - v ({{< req >}})
+- Returns a value the same type as the two input parameters
+
+```js
+(?multiplier: A, v: A) => A
+```
+
+---
+
+#### Pass-through transformation
+The following type signature describes a
+[transformation](/flux/v0.x/function-types/#transformations) that:
+
+- Takes a stream of tables as piped-forward input
+- Returns the input stream of tables unmodified
+
+```js
+(<-tables: stream[A]) => stream[A]
+```
+
+---
+
+#### Basic transformation
+The following type signature describes a
+[transformation](/flux/v0.x/function-types/#transformations) that:
+
+- Takes a stream of tables as piped-forward input
+- Returns a new, modified stream of tables
+
+```js
+(<-tables: stream[A]) => stream[B]
```

Comment:
This is a discouraged pattern in Flux. However only built-ins can do this. We call this _return type polymorphism_ and its generally bad. Notice how `B` only shows up on the return side of the type but not the input side? This means Flux cannot connect the output type to the input types in any way. Which means that Flux does not know anything about the return type, and so therefore can no longer provide the user with useful error messages. Most languages do not even allow for this kind of thing. Flux does in order to be backwards compatible with its pre-typed days. 

So this is example is accurate and it does occur in the wild, however the Flux team is actively trying to reduce/eliminate these cases. 


Perhaps a better although slightly more complex example would be map?

```
builtin map : (<-tables: stream[A], fn: (r: A) => B,) => stream[B]
```

Notice how its very similar in that you have a `stream[A]` as input and `stream[B]` as output. But this time we also have `fn` which tells us how `A` relates to `B`.  Since `B` occurs on the input side with this signature Flux can determine exactly what `B` is and therefore we can continue to provide useful error messages/hints to users.

Happy to chat about this more if you want more details.